operator/controllers/execution/scans/hook_reconciler.go

```diff
@@ -62,250 +60,232 @@ func (r *ScanReconciler) setHookStatus(scan *executionv1.Scan) error {
 	return nil
 }
 
-func (r *ScanReconciler) executeReadAndWriteHooks(scan *executionv1.Scan) error {
-	// Get the first Hook Status which is not completed.
-	ctx := context.Background()
-	var nonCompletedHook *executionv1.HookStatus
-
-	for _, hook := range scan.Status.ReadAndWriteHookStatus {
+func getNonCompletedHookPriorityQueue(hooks *[]*executionv1.HookStatus) util.PriorityQueue {
+	var priorityQueue = make(util.PriorityQueue, len(*hooks))
+	var completedHooks = 0
+	for _, hook := range *hooks {
 		if hook.State != executionv1.Completed {
-			nonCompletedHook = &hook
-			break
+			priorityQueueItem := util.PriorityQueueItem{
+				Value:    hook,
+				Priority: hook.Priority,
+			}
+			priorityQueue[completedHooks] = &priorityQueueItem
+			completedHooks++
 		}
 	}
+	priorityQueue = priorityQueue[:completedHooks]
+	heap.Init(&priorityQueue)
+	return priorityQueue
+}
+
+func (r *ScanReconciler) executeReadAndWriteHooks(scan *executionv1.Scan) error {
+	ctx := context.Background()
 
+	nonCompletedHooks := getNonCompletedHookPriorityQueue(&scan.Status.ReadAndWriteHookStatus)
+
+	var err error
 	// If nil then all hooks are done
-	if nonCompletedHook == nil {
-		scan.Status.State = "ReadAndWriteHookCompleted"
-		if err := r.Status().Update(ctx, scan); err != nil {
-			r.Log.Error(err, "unable to update Scan status")
-			return err
+	if len(nonCompletedHooks) == 0 {
+		r.Log.Info("All ReadAndWriteHooks have been completed", "ScanName", scan.Name)
+		scan.Status.State = "ReadOnlyHookProcessing"
+	} else if len(nonCompletedHooks) > 0 {
+		priorityQueueItem := heap.Pop(&nonCompletedHooks).(*util.PriorityQueueItem)
+		hook := priorityQueueItem.Value.(*executionv1.HookStatus)
+
+		err = r.processHook(scan, hook)
+		if err != nil || hook.State == executionv1.Failed || hook.State == executionv1.Cancelled {
+			scan.Status.State = "Errored"
+			scan.Status.ErrorDescription = fmt.Sprintf("Failed to execute ReadAndWrite Hook '%s' in job '%s'. Check the logs of the hook for more information.", hook.HookName, hook.JobName)
 		}
-		return nil
 	}
 
-	switch nonCompletedHook.State {
-	case executionv1.Pending:
-		rawFileURL, err := r.PresignedGetURL(scan.UID, scan.Status.RawResultFile, defaultPresignDuration)
-		if err != nil {
-			return err
-		}
-		findingsFileURL, err := r.PresignedGetURL(scan.UID, "findings.json", defaultPresignDuration)
-		if err != nil {
-			return err
+	if err == nil && scan.Status.State == "ReadAndWriteHookProcessing" {
```

Comment:
Can you compare to constants (outside this module) rather than literals?

Reply:
This isn't introduced in this PR but something we should definitely do 👍